### PR TITLE
Shut down the thread for tactic phase properly

### PR DESCRIPTION
**Pull request by Adriankhl**
## Issue

1. Use a hero with the tactic skill to enter a battle
2. Load game during the tactic phase
3. Crash

## Fix
The thread for tactic phase keeps running even if the game is ended. This PR shut that down properly.

This might not be the most elegant solution since I need to create a `forceEndTacticPhase()` function. However, I couldn't figure out a good solution given how the thread is implemented. 

Btw, I have a feeling that it is very easy to make mistake with threads in VCMI, I suggest that people who review PR (including mine :smile:) with threading-related code should pay extra attention to ensure that they work properly.

### CODE REVIEW

**Comment by krs0** [comment]
@Adriankhl VCMI developers use Slack for communication. If you want to join...
https://slack.vcmi.eu/

File: client/Client.cpp [hunk 1]
```diff
@@ -593,7 +598,7 @@ void CClient::battleStarted(const BattleInfo * info)
 
 	if(info->tacticDistance && vstd::contains(battleints, info->sides[info->tacticsSide].color))
 	{
-		boost::thread(&CClient::commenceTacticPhaseForInt, this, battleints[info->sides[info->tacticsSide].color]);
+		playerTacticThreads[info->sides[info->tacticsSide].color] = std::make_unique<boost::thread>(boost::thread(&CClient::commenceTacticPhaseForInt, this, battleints[info->sides[info->tacticsSide].color]));
```
Comment (nullkiller) [inline client/Client.cpp:601]: Do we need to specify it once again here if we already said it in std::make_unique<boost::thread>? I am not c++ pro, just want to ensure there will be no copy-construction here

Comment (nullkiller) [reply client/Client.cpp:601]: std::make_unique<boost::thread>(boost::thread(&CClient::commenceTacticPhaseForInt, this, battleints[info->sides[info->tacticsSide].color]));
```suggestion
		playerTacticThreads[info->sides[info->tacticsSide].color] = std::make_unique<boost::thread>(&CClient::commenceTacticPhaseForInt, this, battleints[info->sides[info->tacticsSide].color]);

```

Comment (IvanSavenko) [reply client/Client.cpp:601]: I believe boost::thread is already non-copyable, so copy-construction should be impossible.
However this line is very overloaded. Perhaps split it?
```suggestion
		PlayerColor color = info->sides[info->tacticsSide].color;
		playerTacticThreads[color] = std::make_unique<boost::thread>(&CClient::commenceTacticPhaseForInt, this, battleints[color]);

```

Comment (Nordsoft91) [inline client/Client.cpp:601]: does it make sense to check if playerTacticThreads[info->sides[info->tacticsSide].color] thread is finished before creating a new thread?

Comment (Adriankhl) [reply client/Client.cpp]: Yes this looks better

Comment (Adriankhl) [reply client/Client.cpp]: How can I check this? Call join() if the thread exists?

File: client/Client.cpp [hunk 2]
```diff
@@ -593,7 +598,8 @@ void CClient::battleStarted(const BattleInfo * info)
 
 	if(info->tacticDistance && vstd::contains(battleints, info->sides[info->tacticsSide].color))
 	{
-		boost::thread(&CClient::commenceTacticPhaseForInt, this, battleints[info->sides[info->tacticsSide].color]);
+		PlayerColor color = info->sides[info->tacticsSide].color;
```
Comment (Nordsoft91) [inline client/Client.cpp:601]: can be moved out: 
`if(info->tacticDistance && vstd::contains(battleints, info->sides[info->tacticsSide].color))`
replace with
```
PlayerColor color = info->sides[info->tacticsSide].color;
if(info->tacticDistance && vstd::contains(battleints, color))
```

**Review by Nordsoft91 — APPROVED** [review]

File: client/Client.cpp (hunk 2 above)
Comment (Adriankhl) [reply client/Client.cpp:601]: There are many `color` inside that function (`battleStarted`), moving the `color` out might confuse other people in the future, I think there is a reason to keep this inside the if bracket.

**Pull request merged**